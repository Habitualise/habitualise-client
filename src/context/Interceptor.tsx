import {createContext, useState, useEffect, useMemo} from 'react';
import {useAuth0} from 'react-native-auth0';
import {axios} from '../lib/axios';

export const InterceptorContext = createContext({isTokenSet: false});

const Interceptor = ({children}) => {
  const {loading, getCredentials} = useAuth0();
  const [accessToken, setAccessToken] = useState();

  useEffect(() => {
    const getAccessToken = async () => {
      try {
        const token = await getCredentials();
        if (token && token.accessToken) {
          setAccessToken(token.accessToken);
        }
      } catch (error) {
        console.log(error);
      }
    };
    if (!loading) {
      getAccessToken();
    }
  }, [loading, getCredentials]);

  useMemo(() => {
    axios.interceptors.request.use(
      async config => {
        const getAccessToken = async () => {
          try {
            const token = await getCredentials();
            if (token && token.accessToken) {
              setAccessToken(token.accessToken);
            }
          } catch (error) {
            console.log(error);
          }
        };
        if (!loading) {
          getAccessToken();
        }
        if (accessToken) {
          console.log(
            '=== setting axios interceptor accessToken Interceptor.tsx [44] ===',
            accessToken,
          );
          config.headers.Authorization = `Bearer ${accessToken}`;
        }
        return config;
      },
      error => {
        return Promise.reject(error);
      },
    );
  }, [loading, getCredentials, accessToken]);

  return (
    <InterceptorContext.Provider value={{isTokenSet: !!accessToken}}>
      {children}
    </InterceptorContext.Provider>
  );
};

export default Interceptor;
