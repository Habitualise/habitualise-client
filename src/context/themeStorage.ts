import {MMKVLoader} from 'react-native-mmkv-storage';

const themeStorage = new MMKVLoader().withInstanceID('theme').initialize();

export const setTheme = (isDark: boolean) => {
  themeStorage.setString('theme', isDark ? 'dark' : 'light');
};

export const getTheme = () => {
  const theme = themeStorage.getString('theme');
  return theme === 'dark';
};

export default themeStorage;
