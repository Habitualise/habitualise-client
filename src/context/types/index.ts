export interface Habit {
  id: string;
  name: string;
  iconName: string;
  colour: HabitColor;
  isCompleted: boolean;
  completionPercentage: number;
  active: boolean;
  daysDue: number[];
  completionHistory: boolean[]; // DONE, SKIPPED, MISSED - one for each day, last element is the most recent (today)
}

export type HabitColor =
  | 'red'
  | 'green'
  | 'blue'
  | 'yellow'
  | 'purple'
  | 'orange'
  | 'grey';

export interface UserBE {
  name: string;
}
