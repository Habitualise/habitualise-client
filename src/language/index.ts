export const LABEL = {
  WELCOME_HEADER: 'Ready to\nbecome giga?',
  WELCOME_SUBHEADER: 'Start your journey with us\ntoday on project habits.',
  SIGN_UP: 'Sign Up',
  ALREADY_HAVE_ACCOUNT: 'I already have an account',
  LOGIN_SCREEN: 'Login Screen',
  LOG_IN: 'Log In',
  LOG_OUT: 'Log Out',
  WELCOME_SCREEN: 'Welcome Screen',
  TODAY_SCREEN: 'Today Screen',
  TODAY_HEADER: "Today's Habits",
  TODAY: 'Today',
  HABITS_SCREEN: 'Habits Screen',
  HABITS_HEADER: 'My Habits',
  ACTIVE: 'Active',
  ARCHIVED: 'Archived',
  HABITS: 'Habits',
  GROUPS_SCREEN: 'Groups Screen',
  GROUPS: 'Groups',
  SETTINGS_SCREEN: 'Settings Screen',
  SETTINGS: 'Settings',
  LOGGED_IN_AS: (username: string) => `Logged in as ${username}`,
  DAYS_COMPLETED: (percentage: number) => `${percentage}% of days completed`,
  ADD_HABIT: 'Add a new habit',
  HABIT_NAME: 'Habit Name',
  REPEAT: 'REPEAT',
};
