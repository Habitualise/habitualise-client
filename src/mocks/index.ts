export const mockUser = {
  email: 'lokchoy@gmail.com',
  email_verified: true,
  sub: 'auth0|5f2e6f7e6e1c6f0c6c0d6b7d',
};

export const mockHabits = [
  {
    id: 1,
    name: 'Running',
    iconName: 'run',
    colour: 'rgb(66, 135, 245)',
    isCompleted: false,
    completionPercentage: 67,
    status: 'active',
    daysDue: [1, 2, 3, 4, 5, 6],
    lastSevenCompletions: ['2023-02-01', '2023-02-02', '2023-02-03'],
  },
  {
    id: 2,
    name: 'Yoga',
    iconName: 'yoga',
    colour: 'rgb(245, 179, 66)',
    isCompleted: false,
    completionPercentage: 0,
    status: 'active',
    daysDue: [1, 2, 3, 4],
    lastSevenCompletions: [],
  },
  {
    id: 3,
    name: 'Meditation',
    iconName: 'meditation',
    colour: 'rgb(207, 65, 250)',
    isCompleted: false,
    completionPercentage: 99,
    status: 'active',
    daysDue: [0, 1, 2, 3, 4, 5, 6],
    lastSevenCompletions: [
      '2023-02-01',
      '2023-02-02',
      '2023-02-07',
      '2023-02-13',
      '2023-02-15',
      '2023-02-16',
    ],
  },
];
