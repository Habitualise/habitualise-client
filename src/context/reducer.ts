import {IState} from '@/types';

export const ACTIONS = {
  SET_ALL_HABITS: 'SET_ALL_HABITS',
  TOGGLE_HABIT_IS_COMPLETED: 'TOGGLE_HABIT_IS_COMPLETED',
  HANDLE_LOGOUT: 'HANDLE_LOGOUT',
};

export const reducer = (state: IState, action) => {
  switch (action.type) {
    case ACTIONS.SET_ALL_HABITS:
      // action.payload should be an array of habits
      // TODO: add some validation here
      return {...state, habits: action.payload};
    case ACTIONS.HANDLE_LOGOUT:
      return {...state, habits: []};
    case ACTIONS.TOGGLE_HABIT_IS_COMPLETED: {
      // action.payload should be the id of the habit
      // scan through habits and find the one that matches the id
      // then toggle the isCompleted property
      const updatedHabits = state.habits.map(habit => {
        if (habit.id === action.payload) {
          const updatedIsCompleted = !habit.isCompleted;
          // today's date e.g. '2023-02-13' to add to or remove from lastSevenCompletions array
          const today = new Date();
          const todayString = today.toISOString().split('T')[0];

          if (updatedIsCompleted) {
            habit.lastSevenCompletions.push(todayString);
          } else {
            habit.lastSevenCompletions = habit.lastSevenCompletions.filter(
              date => date !== todayString,
            );
          }

          return {...habit, isCompleted: !habit.isCompleted};
        }
        return habit;
      });
      return {...state, habits: updatedHabits};
    }
    default:
      return state;
  }
};
