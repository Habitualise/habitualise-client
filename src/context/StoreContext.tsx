import * as React from 'react';
import {State} from './reducer';

export * from './reducer';

interface HabitsStore {
  state: State;
  dispatch: React.Dispatch<any>;
}

const StoreContext = React.createContext<HabitsStore>({
  state: {habits: []},
  dispatch: () => {},
});

const initState = {
  habits: [],
};

interface StoreContextProviderProps {
  children: React.ReactNode;
  reducer: React.Reducer<any, any>;
}

export const StoreContextProvider = ({
  children,
  reducer,
}: StoreContextProviderProps) => {
  const [state, dispatch] = React.useReducer(reducer, initState);
  const memoizedValue = React.useMemo(() => ({state, dispatch}), [state]);
  return (
    <StoreContext.Provider value={memoizedValue}>
      {children}
    </StoreContext.Provider>
  );
};

export const useStore = () => React.useContext(StoreContext);
