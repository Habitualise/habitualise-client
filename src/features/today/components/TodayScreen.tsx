import React from 'react';
import {ScrollView, View} from 'react-native';
import {Divider, Text} from 'react-native-paper';
import {SafeAreaView} from 'react-native-safe-area-context';

import {LABEL} from '@app/language';
import getTodayDateString from '../utils/getTodayDateString';
import {TodayHabit} from './TodayHabit';
import {useStore} from '@app/context/StoreContext';
import PaperView from '@app/components/PaperView';
import {commonStyles} from '@app/components/styles';
import {getTodaysHabits} from '@app/features/today/utils/getTodaysHabits';
import Spinner from '@app/components/Spinner';

export const TodayScreen = () => {
  const {state} = useStore();
  const {loading} = state;
  const allHabits = state.habits;
  const todaysHabits = getTodaysHabits(allHabits);

  return (
    <SafeAreaView
      style={commonStyles.safeArea}
      edges={['top', 'left', 'right']}>
      <PaperView style={{flex: 1}}>
        <View style={commonStyles.headerContainer}>
          <Text variant="bodySmall">{getTodayDateString()}</Text>
          <Text variant="headlineLarge" style={commonStyles.heading}>
            {LABEL.TODAY_HEADER}
          </Text>
        </View>
        <Divider />
        <ScrollView>
          {loading && <Spinner />}
          {todaysHabits.map(habit => (
            <TodayHabit
              key={habit.id}
              id={habit.id}
              iconName={habit.iconName}
              name={habit.name}
              isCompletedToday={habit.isCompletedToday}
              daysDue={habit.daysDue}
              colour={habit.colour}
            />
          ))}
        </ScrollView>
      </PaperView>
    </SafeAreaView>
  );
};
