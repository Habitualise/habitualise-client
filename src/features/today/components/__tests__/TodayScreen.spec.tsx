import React from 'react';
import { render } from '@testing-library/react-native';

import { TodayScreen } from '../TodayScreen';
import { LABEL } from '@/language';
import { reducer, StoreContextProvider } from '@/context/StoreContext';
import { mockHabits } from '@/mocks';

// mock useStore state
jest.mock('@/context/StoreContext', () => ({
  ...jest.requireActual('@/context/StoreContext'),
  useStore: () => ({
    state: {
      habits: mockHabits,
    },
  }),
}));

// mock toLocaleDateString to return a consistent value
jest.mock('../../utils/getTodayDateString', () => ({
  __esModule: true,
  default: () => 'Saturday, 11 February',
}));

describe('TodayScreen', () => {
  it('should match snapshot', () => {
    const { toJSON } = render(
      <StoreContextProvider reducer={reducer}>
        <TodayScreen />
      </StoreContextProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should display todays habits header', () => {
    const { getByText } = render(
      <StoreContextProvider reducer={reducer}>
        <TodayScreen />
      </StoreContextProvider>,
    );
    expect(getByText(LABEL.TODAY_HEADER)).toBeTruthy();
  });
});
