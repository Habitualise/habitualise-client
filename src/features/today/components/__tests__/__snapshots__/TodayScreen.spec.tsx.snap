// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodayScreen should match snapshot 1`] = `
<RNCSafeAreaView
  edges={
    [
      "top",
      "left",
      "right",
    ]
  }
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(255, 251, 254, 1)",
        },
        {
          "flex": 1,
        },
      ]
    }
  >
    <View
      style={
        {
          "paddingHorizontal": 16,
          "paddingVertical": 24,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontSize": 12,
              "fontWeight": "400",
              "letterSpacing": 0.4,
              "lineHeight": 16,
            },
            {
              "textAlign": "left",
            },
            {
              "writingDirection": "ltr",
            },
            undefined,
          ]
        }
      >
        Saturday, 11 February
      </Text>
      <Text
        style={
          [
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontSize": 32,
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 40,
            },
            {
              "textAlign": "left",
            },
            {
              "writingDirection": "ltr",
            },
            {
              "fontWeight": "600",
            },
          ]
        }
      >
        Today's Habits
      </Text>
    </View>
    <View
      style={
        [
          {
            "backgroundColor": "rgba(202, 196, 208, 1)",
            "height": 0.5,
          },
          undefined,
          false,
          false,
          undefined,
        ]
      }
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "backgroundColor": "rgba(0,0,0,0)",
          },
        ]
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(0,0,0,0)",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "padding": 16,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(66, 135, 245, 0.15)",
              "borderRadius": 50,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginRight": 16,
              "padding": 5,
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "rgb(66, 135, 245)",
                  "fontSize": 32,
                },
                undefined,
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󰜎
          </Text>
        </View>
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": "rgb(33, 33, 33)",
                  "fontWeight": "bold",
                  "marginBottom": 4,
                  "textDecorationLine": "none",
                },
              ]
            }
          >
            Running
          </Text>
          <Text
            style={
              [
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontSize": 12,
                  "fontWeight": "400",
                  "letterSpacing": 0.4,
                  "lineHeight": 16,
                },
                {
                  "textAlign": "left",
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": "rgb(158, 158, 158)",
                },
              ]
            }
          >
            Every day
          </Text>
        </View>
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "rgb(189, 189, 189)",
                "fontSize": 25,
              },
              undefined,
              {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          󰝦
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "rgba(202, 196, 208, 1)",
              "height": 0.5,
            },
            undefined,
            false,
            false,
            undefined,
          ]
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "backgroundColor": "rgba(0,0,0,0)",
          },
        ]
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(0,0,0,0)",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "padding": 16,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(245, 179, 66, 0.15)",
              "borderRadius": 50,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginRight": 16,
              "padding": 5,
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "rgb(245, 179, 66)",
                  "fontSize": 32,
                },
                undefined,
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󱅼
          </Text>
        </View>
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": "rgb(33, 33, 33)",
                  "fontWeight": "bold",
                  "marginBottom": 4,
                  "textDecorationLine": "none",
                },
              ]
            }
          >
            Yoga
          </Text>
          <Text
            style={
              [
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontSize": 12,
                  "fontWeight": "400",
                  "letterSpacing": 0.4,
                  "lineHeight": 16,
                },
                {
                  "textAlign": "left",
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": "rgb(158, 158, 158)",
                },
              ]
            }
          >
            Mon, Tue, Wed, Thu, Sun
          </Text>
        </View>
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "rgb(189, 189, 189)",
                "fontSize": 25,
              },
              undefined,
              {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          󰝦
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "rgba(202, 196, 208, 1)",
              "height": 0.5,
            },
            undefined,
            false,
            false,
            undefined,
          ]
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "backgroundColor": "rgba(0,0,0,0)",
          },
        ]
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(0,0,0,0)",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "padding": 16,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(207, 65, 250, 0.15)",
              "borderRadius": 50,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginRight": 16,
              "padding": 5,
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "rgb(207, 65, 250)",
                  "fontSize": 32,
                },
                undefined,
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󱅻
          </Text>
        </View>
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": "rgb(33, 33, 33)",
                  "fontWeight": "bold",
                  "marginBottom": 4,
                  "textDecorationLine": "none",
                },
              ]
            }
          >
            Meditation
          </Text>
          <Text
            style={
              [
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontSize": 12,
                  "fontWeight": "400",
                  "letterSpacing": 0.4,
                  "lineHeight": 16,
                },
                {
                  "textAlign": "left",
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": "rgb(158, 158, 158)",
                },
              ]
            }
          >
            Every day
          </Text>
        </View>
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "rgb(189, 189, 189)",
                "fontSize": 25,
              },
              undefined,
              {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          󰝦
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "rgba(202, 196, 208, 1)",
              "height": 0.5,
            },
            undefined,
            false,
            false,
            undefined,
          ]
        }
      />
    </View>
  </View>
</RNCSafeAreaView>
`;
